### PR TITLE
added memory usage utility

### PR DESCRIPTION
**Pull request by michaeljbrazell**

### CODE REVIEW

File: src/OversetSimulation.cpp [hunk 1]
```diff
@@ -143,4 +147,30 @@ bool OversetSimulation::do_connectivity(const int tstep)
     return (tstep > 0) && (tstep % m_overset_update_interval) == 0;
 }
 
+long OversetSimulation::mem_usage(bool writefile)
+{
+    struct rusage usage;
+    int i;
+    getrusage(RUSAGE_SELF, &usage);
+    /* convert to MB */
```
Comment (marchdf) [inline src/OversetSimulation.cpp:155]: can we convert comments to just `//` ?

File: src/OversetSimulation.cpp [hunk 2]
```diff
@@ -143,4 +147,30 @@ bool OversetSimulation::do_connectivity(const int tstep)
...
+    long mem = (long) ((double) usage.ru_maxrss)/1024.0;
+
+    int psize, prank;
+    MPI_Comm_size(m_comm, &psize);
+    MPI_Comm_rank(m_comm, &prank);
+
+    /* gather all memory usage to proc 0 */
+    long memall[psize];
```
Comment (marchdf) [inline src/OversetSimulation.cpp:163]: can we use `vector` instead of C arrays?

File: src/OversetSimulation.cpp [hunk 3]
```diff
@@ -143,4 +147,30 @@ bool OversetSimulation::do_connectivity(const int tstep)
...
+    MPI_Gather(&mem,1,MPI_LONG,&memall,1,MPI_LONG,0,m_comm);
+    if (prank == 0 && writefile) {
+        FILE *fp;
+        char filename[] = "memusage.dat";
```
Comment (marchdf) [inline src/OversetSimulation.cpp:167]: replace with `std::string`?

File: src/OversetSimulation.cpp [hunk 4]
```diff
@@ -143,4 +147,30 @@ bool OversetSimulation::do_connectivity(const int tstep)
...
+        fp = fopen(filename,"a");
```
Comment (marchdf) [inline src/OversetSimulation.cpp:168]: Use a more c++ way for outputting to a file?

File: src/ExawindSolver.h [hunk 5]
```diff
@@ -77,13 +79,50 @@ class ExawindSolver
...
         printer.echo(out + " " + timings);
     }
+
+    long mem_usage()
+    {
+        struct rusage usage;
+        getrusage(RUSAGE_SELF, &usage);
+        
+        // convert to MB
+        long mem = (long) ((double) usage.ru_maxrss)/1024.0;
```
Comment (marchdf) [inline src/ExawindSolver.h:97]: `const` this?

File: src/OversetSimulation.cpp [hunk 6]
```diff
@@ -143,4 +149,46 @@ bool OversetSimulation::do_connectivity(const int tstep)
     return (tstep > 0) && (tstep % m_overset_update_interval) == 0;
 }
 
+long OversetSimulation::mem_usage_all(const int step)
+{
+    struct rusage usage;
+    int i;
+    getrusage(RUSAGE_SELF, &usage);
+    // convert to MB
+    long mem = (long) ((double) usage.ru_maxrss)/1024.0;
```
Comment (marchdf) [inline src/OversetSimulation.cpp:158]: const this?

File: src/OversetSimulation.cpp [hunk 7]
```diff
@@ -1,5 +1,7 @@
 #include "OversetSimulation.h"
 #include <algorithm>
+#include <sys/resource.h>
```
Comment (marchdf) [inline src/OversetSimulation.cpp:3]: can we ifdef this stuff for linux?

File: src/ExawindSolver.h [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 #include "Timers.h"
 #include "ParallelPrinter.h"
+#include <sys/resource.h>
```
Comment (marchdf) [inline src/ExawindSolver.h:6]: can we ifdef this stuff for linux?

File: src/OversetSimulation.cpp [hunk 9]
```diff
@@ -143,4 +149,45 @@ bool OversetSimulation::do_connectivity(const int tstep)
     return (tstep > 0) && (tstep % m_overset_update_interval) == 0;
 }
 
+long OversetSimulation::mem_usage_all(const int step)
+{
+    struct rusage usage;
```
Comment (marchdf) [inline src/OversetSimulation.cpp:154]: This has a lot in common with the other call in ExawindSolver. Can we maybe consolidate the memory stuff in a separate file? and then put the ifdefs there? That way we we ever want to update the memory stuff for other platforms we can just look at that file.

**Review by marchdf — CHANGES_REQUESTED** [review]

**Review by marchdf — APPROVED** [review]

**Comment by marchdf** [comment]
This is super cool and we can merge if Mike is happy with my edits!

**Comment by michaeljbrazell** [comment]
yeah that looks good!

**Pull request merged**